src/distributions/uniform.rs

```diff
@@ -534,59 +533,110 @@ macro_rules! uniform_float_impl {
 
             fn new(low: Self::X, high: Self::X) -> Self {
                 assert!(low < high, "Uniform::new called with `low >= high`");
-                let scale = high - low;
-                let offset = low - scale;
-                UniformFloat {
-                    scale: scale,
-                    offset: offset,
+                assert!(low.is_finite() && high.is_finite(), "Uniform::new called with non-finite boundaries");
+
+                let mut scale = high - low;
+
+                let max_rand = (::core::$uty::MAX >> $bits_to_discard)
+                               .into_float_with_exponent(0) - 1.0;
+
+                while scale * max_rand + low >= high {
+                    scale = <$ty>::from_bits(scale.to_bits() - 1);
                 }
+
+                debug_assert!(scale >= 0.0);
```

Comment:
If `high` is the next representable number above `low` then `scale` is effectively zero — although larger values than zero may still give the same rounding. Probably though we should accept zero.

Note that your `while` loop would take a long time to get to zero since `high - low` may be much larger than the minimal representable value, though it likely doesn't matter since it should encounter a value of `scale` small enough that `low + scale` rounds to `low` much sooner.

Reply:
Yeah, I'm banking on that `scale * max_rand + low > high` only happens in situations with unfortunate rounding, so we should only need to make small adjustments to `scale` to get it within the acceptable range. Especially since `max_rand` is already < 1.